### ISSUE TITLE
[CGPO] Add support for Constrained Generative Policy Optimization

### ISSUE THREAD

**Issue by gaetanlop** [huggingface/trl#2156]
### Method description

**Feature request**

Constrained Generative Policy Optimization was introduced by Meta in a recent paper (https://arxiv.org/pdf/2409.20370). It seems to outperform PPO and DPO and is specifically designed to address standard RLHF limitations (reward hacking, accurate alignment guidance...). 

They introduced 2 options for rlhf optimizers:  1. `CalibratedRegularized Policy Gradient` (CRPG) and  `Constrained Online DPO` (CODPO). 2. `Calibrated-Regularized Reward Ranking Fine-tuning` (CRRAFT).

The implementation will likely require the following steps:

- [ ] Calibrated reward modeling (#2155)
- [ ] A Mixture of judges
- [ ] CGPO parent class that defines core methods based on the algorithm described in section 4.1 of the paper. 

Specific subclasses that inherit from CGPO and implement different optimizers:
- [ ] CRPG 
- [ ] CODPO (similar to Online DPO)
- [ ] CRRAFT (similar to RAFT)

I suppose HF is interested to support CGPO but just to be sure, Wdyt @lewtun @kashif ? Any interest in supporting CGPO?

**Your contribution**

I can work on this. I have already started in #2155. Next steps might be to add the mixture of judges and the CGPO parent class.

### Open source status

- [ ] The method implementation is available
- [ ] The model weights are available
- [ ] The training datasets are available

**Comment by lewtun** [comment]
Hey @gaetanlop I think this would be a really cool addition to the library and one that is likely to be useful to the community now that online methods are becoming more common!

The best part is that they used open datasets in the paper, so we should be able to verify the implementation by trying to reproduce similar training curves like this:

<img width="1208" alt="Screenshot 2024-10-02 at 18 18 46" src="https://github.com/user-attachments/assets/22dbed91-e5ff-4e25-aeef-b600a560e0fa">

I suggest starting with small models from Llama or Qwen2.5 and later we can scale up on the HF cluster if needed :)

**Comment by gaetanlop** [comment]
Hello @lewtun, thank you for your reply. Indeed, we should be able to verify that everything is correct by trying to reproduce there results. The judges seem to be using pretty large models though (llama 70b), so this might be complicated to run it without your cluster.

I have implemented the reward part and the mixture of judges in separate PRs, the rest will be done in a single PR normally. I can also close the 2 other PRs and put everything in a single PR if you prefer.

**Comment by albertvillanova** [comment]
Hey @gaetanlop, sorry for the very late reply — catching up on a long backlog. 

Thanks again for the careful breakdown and for the work you did on the prerequisite pieces. Where things landed: `MixtureOfJudges` from #2159 was merged, but the CGPO trainer itself (#2190) ended up closed for inactivity in mid-2025, and shortly after we removed the entire `trl.experimental.judges` module and all judge support from the online trainers in #5485 — the rationale being no known production usage and the fact that `reward_funcs` (now accepting arbitrary callables) covers the same surface with less dual-code-path maintenance. Since CGPO is fundamentally judge-driven (MoJ, constraint judges) and we've intentionally moved away from that abstraction, I'm going to close this as wont-implement. 

If you (or anyone landing here) wants to take another swing at CGPO expressed purely via `reward_funcs`, happy to discuss the design in a fresh issue.
